python-sdk/dev/scripts/replace_dependencies.py

```diff
@@ -0,0 +1,110 @@
+from __future__ import annotations
+
+import argparse
+import fileinput
+import pathlib
+from re import sub
+from urllib.parse import urlparse
+
+import requests
+from bs4 import BeautifulSoup
+from pip._internal.utils.packaging import get_requirement
+
+ROOT = pathlib.Path(__file__).parent.parent.parent
+
+
+def _parse_pinned_pacakge_from_cncf(cncf_url: str) -> str:
+    """Parse the pinned version from apache-airflow-providers-cncf-kubernetes Download files page [1]
+    As we're not able to test apache-airflow-providers-cncf-kubernetes with pinned version directly,
+    we'll need to install from the Source Distribution [2].
+    Take input cncf_url=https://pypi.org/project/apache-airflow-providers-cncf-kubernetes/6.2.0rc1/
+    as an example, this function goes to Download files page and retreieve Source Distribution URL[2]
+    [1] https://pypi.org/project/apache-airflow-providers-cncf-kubernetes/6.2.0rc1/#files
+    [2] https://files.pythonhosted.org/packages/08/0f/0cf8e0895995055edc5df9575c5cb32debbf25eb7019bb70dc6d14f36e0d/apache-airflow-providers-cncf-kubernetes-6.2.0rc1.tar.gz
+    :param cncf_url: PyPI URL of the apache-airflow-providers-cncf-kubernetes with the version to test
+    """
+    download_package_url = f"{cncf_url}#files"
+    req = requests.get(download_package_url)
+    soup = BeautifulSoup(req.text, "html.parser")
+    first_file_card_comment = soup.find("div", {"class": "file__card"})
+    url = first_file_card_comment.a.get("href")
+    return f"apache-airflow-providers-cncf-kubernetes @{url}"
+
+
+def _parse_pinned_package_from_pypi_url(url: str) -> str:
+    """Parse the pinned version from PyPI URL
+    Take input url=https://pypi.org/project/apache-airflow-providers-apache-livy/3.5.0rc1/
+    as an example, this function extract the path of it and returns
+    "apache-airflow-providers-apache-livy==3.5.0rc1"
+    :param cncf_url: PyPI URL of the provider to test
+    """
+    package_name, version = urlparse(url).path.split("/")[2:]
+    return f"{package_name}=={version}"
+
+
+def _parse_pypi_url_from_h2_title(h2_soup: BeautifulSoup) -> str:
+    """Parse the h2 title for retrieving package PyPI urls
+    :param h2_soup: the BeautifulSoup object that contains the h2 title with PyPI urls
+    """
+    return h2_soup.a.get("href")
+
+
+def parse_providers_release_testing_gh_issue(issue_url: str) -> list[str]:
+    """Parse the pinned packages from The URL of the github issue that announce provider testing
+    (e.g., https://github.com/apache/airflow/issues/31322)
+    :param issue_url: The URL of the github issue that announce provider testing
+    """
+    req = requests.get(issue_url)
+    soup = BeautifulSoup(req.text, "html.parser")
```

Comment:
Did we consider the possibility of using the Gihtub HTTP API to retrieve the information we need?